eclair-core/src/main/scala/fr/acinq/eclair/db/psql/PsqlUtils.scala

```diff
@@ -0,0 +1,141 @@
+    statement.executeUpdate("CREATE TABLE IF NOT EXISTS versions (db_name TEXT NOT NULL PRIMARY KEY, version INTEGER NOT NULL)")
+    // overwrite the existing version
+    statement.executeUpdate(s"UPDATE versions SET version=$newVersion WHERE db_name='$db_name'")
+  }
+
+  val LockTable = "lock_leases"
+
+  case class LockLease(expiresAt: Timestamp, instance: String, expired: Boolean)
+
+  class TooManyLockAttempts(msg: String) extends RuntimeException(msg)
+
+  class UninitializedLockTable(msg: String) extends RuntimeException(msg)
+
+  class DatabaseLocked(msg: String) extends RuntimeException(msg)
+
+  def obtainExclusiveLock(instanceId: String, leaseDuration: FiniteDuration, lockTimeout: FiniteDuration, attempt: Int = 1)(implicit ds: DataSource): Unit = synchronized {
+    logger.info(s"Trying to acquire a database lock (attempt #$attempt) instance ID=$instanceId")
+
+    if (attempt > 3) throw new TooManyLockAttempts("Too many attempts to acquire a database lock")
+
+    try {
+      withConnection { implicit connection =>
+        val autocommit = connection.getAutoCommit
+        try {
+          connection.setAutoCommit(false)
+          acquireExclusiveTableLock(lockTimeout)
+          val lease = currentLease
+          if (lease.instance == instanceId) {
+            updateLease(instanceId, leaseDuration)
+          } else if (lease.expired)
+            updateLease(instanceId, leaseDuration, startNew = true)
+          else
+            throw new DatabaseLocked(s"The database is locked by instance ID=${lease.instance}")
+          connection.commit()
+          logger.info("Database lock was successfully acquired.")
+        } catch {
+          case e: Throwable =>
+            connection.rollback()
+            throw e
+        } finally {
+          connection.setAutoCommit(autocommit)
+        }
+      }
+    } catch {
+      case e: PSQLException if (e.getServerErrorMessage != null && e.getServerErrorMessage.getSQLState == "42P01") =>
+        withConnection { connection =>
+          logger.warn(s"Table ${LockTable} does not exist, trying to recreate it...")
+          initializeLocksTable(connection)
+          obtainExclusiveLock(instanceId, leaseDuration, lockTimeout, attempt + 1)
+        }
+    }
+  }
+
+  private def initializeLocksTable(implicit connection: Connection): Unit = {
+    using(connection.createStatement()) { statement =>
+      statement.executeUpdate(s"CREATE TABLE $LockTable (expires_at TIMESTAMP NOT NULL UNIQUE, instance VARCHAR NOT NULL UNIQUE)")
+    }
+  }
+
+  private def acquireExclusiveTableLock(lockTimeout: FiniteDuration)(implicit connection: Connection): Unit = {
+    using(connection.createStatement()) { statement =>
+      statement.executeUpdate(s"SET lock_timeout TO '${lockTimeout.toSeconds}s'")
+      statement.executeUpdate(s"LOCK TABLE $LockTable IN ACCESS EXCLUSIVE MODE")
+    }
+  }
+
+  private def currentLease(implicit connection: Connection): LockLease = {
```

Comment:
`Option[LockLease]` or maybe create an ADT ?

Reply:
It make the logic more complicated.